# Script to generate the data set analysed in Albanese et al. 2023
# source is https://figshare.com/articles/dataset/Input_data_for_systems_genetics_of_sleep_regulation/7797434
# data coming from: Jan, M., Gobet, N., Diessler, S. et al. A multi-omics digital 
# research object for the genetics of sleep regulation. Sci Data 6, 258 (2019).

merge_same_columns <- function(x, y){
  # Helper function to merge the two datasets, since their columns are not in  
  # the same order
  row_names <- c(row.names(x), row.names(y))
  for (dup in row_names[duplicated(row_names)]) {
    row_names[which(row_names==dup)[1]] <- paste(dup, "_t")
  }
  col_names <- c(intersect(names(x), names(y)))
  tmp_df <- rep(0, length(row_names))
  for (name in col_names) {
    assembled_vector <- c(x[, name], y[, name])
    tmp_df <- cbind(tmp_df, assembled_vector)
  }
  tmp_df <- as.data.frame(tmp_df)
  rownames(tmp_df) <- row_names
  tmp_df <- tmp_df[, -1]
  colnames(tmp_df) <- col_names
  tmp_df
}

# Import and format transcriptomic data
# Gene expression file stored in source, in the file bxd.vital-it.ch.backup.zip
t_ctx_sd <- read.table("data-raw/GeneExpression.Cortex.SD.log2CPM.txt.gz", header = T, sep = "\t", row.names = 1)
t_ctx_sd <- t_ctx_sd[, -c(1, 5, 6, 7, seq(38, 43))]
colnames(t_ctx_sd)[4] <- "BXD73b"
colnames(t_ctx_sd)[32] <- "BXD48a"
colnames(t_ctx_sd)[33] <- "BXD65a"
for (i in 1:dim(t_ctx_sd)[2]) {
  name <- colnames(t_ctx_sd)[i]
  if(substring(name, 4, 4)=="0"){
    colnames(t_ctx_sd)[i] <- paste("BXD", substring(name, 5),sep = "")
  }
}

# Import and format metabolomic data
# Metabolite aboundance file is the subset relative to sleep deprivation of the
# file DataSystemsGeneticsOfSleep_mm9.tar.gz/Data/Metabolome/Metabolites_BXD_alldata.txt,
# stored in source
m_sd <- read.csv("data-raw/Metabolites.SD.csv", header = T, row.names = 1)
m_class <- m_sd[, 1]
m_sd <- m_sd[, -c(1,2)]

# Assembling multi-omics data set 
mo_sd <- merge_same_columns(t_ctx_sd, m_sd)
mo_sd <- t(mo_sd)

# Top 100 DEGs from Jan et al. 2019 
# File of differential gene expression stored in source, under
# DataSystemsGeneticsOfSleep_mm9.tar.gz/Data/IntermediateLayer/DE
degs <- read.table("data-raw/RNA_SD_vs_NSD_DiffExp_Limma_Cortex.txt", header = T, row.names = 1)
degs <- degs[degs$adj.P.Val<=0.05,]
degs_100_fc <- row.names(degs[order(abs(degs$logFC), decreasing = T), ])[1:100]

# 78 genes known to be associated with SD from: Mongrain, V. et al. Separating 
# the Contribution of Glucocorticoids and Wakefulness to the Molecular and 
# Electrophysiological Correlates of Sleep Homeostasis. Sleep 33, 1147â€“1157 (2010).
known_genes <- read.table("data-raw/Sleep_genes_Mongrain2010.txt", sep = " ")
# Remove EG665858/EG665787, no entry in current dataset
known_genes <- known_genes[-48,1]

# Update gene synonims to be able to unify the two lists
known_genes[42] <- "Manf"
known_genes[46] <- "Noxred1"
known_genes[56] <- "Pdp1"
known_genes[63] <- "Sik2"
known_genes[73] <- "Mettl16"
genes_subset <- union(degs_100_fc, known_genes)

# Metabolites differentially expressed in Jan et al. 2019
# File of differential metabolite expression was generated by replicating same 
# analysis of the script "Differential_Expression_Metabolites.html" stored in 
# source, in the Source_Code_Systems_Genetics_of_Sleep_Regulation_mm9.zip file 
dems <- read.table("data-raw/Differential_Expression_Metabolites_results.txt",sep="\t", header = T)
dems[,1] <- rownames(m_sd)
dems_sig <- dems[dems$adjPval<=0.05, 1]

# Generate subset of multi-omics data set and 
# remove sample BXD85, it contains 3 NAs
mo_subset <- c(genes_subset, dems_sig)
multi_omics_SD <- mo_sd[-25, mo_subset]

usethis::use_data(multi_omics_SD, overwrite = TRUE)

# Load neighborhood of Cirbp described in Albanese et al. 2023
load("data-raw/cirbp_neighborhood.RData")
multi_omics_SD_small <- multi_omics_SD[, cirbp_neighborhood]

usethis::use_data(multi_omics_SD_small, overwrite = TRUE)
